pkg/controllermanager/controller/shoot/state/finalizer/add.go

```diff
@@ -0,0 +1,76 @@
+// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package finalizer
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+)
+
+// ControllerName is the name of the controller.
+const ControllerName = "shootstate-finalizer"
+
+// AddToManager adds the Reconciler to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+
+	return builder.ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(
+			&gardencorev1beta1.ShootState{},
+			builder.WithPredicates(r.ShootStatePredicates()),
+		).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: ptr.Deref(r.Config.ConcurrentSyncs, 0),
+		}).
+		Watches(
+			&gardencorev1beta1.Shoot{},
+			handler.EnqueueRequestsFromMapFunc(r.MapShootToShootState),
+			builder.WithPredicates(r.ShootPredicates()),
```

Comment:
Same here. However, can't we be "more intelligent" here and only enqueue `ShootState`s for `Shoot`s in migration (i.e., only for relevant `Shoot`s)? Right now, you are enqueueing on every update event.

Reply:
Thanks for pointing to this enhancement, included an additional logic that filters out non-migrating Shoots based on the `.status.LastOperation` transitions. Here's a brief recap of the constraints that currently must be met in order to consider a `Shoot` migration is in progress:

* `.lastOperation.Type` being updated from `Reconcile` to `Migrate`
* `.lastOperation.Type` being updated from `Migrate` to `Restore`
* `.lastOperation` of `Type` `Restore` gets it's `State` updated to `Succesed`
* `.lastOperation.Type` being updated from `Restore` to `Reconcile`